src/track/trackref.cpp

```diff
@@ -0,0 +1,43 @@
+#include "track/trackref.h"
+
+
+bool TrackRef::verifyConsistency() const {
+    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasCanonicalLocation()) {
+        return false;
+    }
+    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasId()) {
+        return false;
+    }
+    return true;
+}
+
+QFileInfo TrackRef::createFileInfo() const {
+    QFileInfo fileInfo(getLocation());
```

Comment:
TrackRef should be a lightweight, almost immutable class with simple 
properties. An instance of QFileInfo can be stored TrackInfoObject if really 
needed.

I currently move all the file-related stuff into another new class 
TrackFile. Then TrackRef simply becomes a composition of TrackFile + TrackId.

After we decided on how to normalize file paths properly to calculate a 
well-defined location for each file, this code might be placed in TrackFile, 
too. It has to take into account the library context like the top-level 
directories which can be provided as an additional parameter. Until then the 
location is calculated as the absolute path of the file, just as before.

On 09/16/2015 10:14 PM, Daniel Schürmann wrote:
>
> In src/track/trackref.cpp 
> <https://github.com/mixxxdj/mixxx/pull/709#discussion_r39678966>:
>
> > @@ -0,0 +1,43 @@
> > +#include "track/trackref.h"
> > +
> > +
> > +bool TrackRef::verifyConsistency() const {
> > +    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasCanonicalLocation()) {
> > +        return false;
> > +    }
> > +    DEBUG_ASSERT_AND_HANDLE(hasLocation() || !hasId()) {
> > +        return false;
> > +    }
> > +    return true;
> > +}
> > +
> > +QFileInfo TrackRef::createFileInfo() const {
> > +    QFileInfo fileInfo(getLocation());
>
> Is this a performance penalty? If yes, does it make sense to cache the 
> object?
> It is probably hard because of concurrency ....
> Thoughts?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/mixxxdj/mixxx/pull/709/files#r39678966>.
>